### ISSUE TITLE
False positive ``unexpected-keyword-arg`` (E1123) for decorators

### ISSUE THREAD

**Issue by pylint-bot** [PyCQA/pylint#258]
Originally reported by: **the mulhern (BitBucket: [the_mulhern](http://bitbucket.org/the_mulhern))**

---

For the following example,  error 
"Passing unexpected keyword argument 'foo' in function call (unexpected-keyword-arg)"
is reported, correctly.

```
#!python

def junk1(junk=None):
    print("%s" % junk)

def main():
    junk1(foo=2)
```

For the example below, no such error is reported, due to the use of keyword args parameter. This is also correct.

```
#!python

def junk1(junk=None, *args, **kwargs):
    print("%s" % junk)

def main():
    junk1(foo=2)
```

In the following example, the decorator supplies the foo argument, but the analysis still reports
"Passing unexpected keyword argument 'foo' in function call (unexpected-keyword-arg)"
which is incorrect. It should be able to detect that foo is now supplied, and give no error.

```
#!python

def decorator(f):

    def new_func(foo=3):
        f(junk=foo)

    return new_func

@decorator
def junk1(junk=None):
    print("%s" % junk)

def main():
    junk1(foo=2)
```

And for the following example it still reports 
"Passing unexpected keyword argument 'foo' in function call (unexpected-keyword-arg)"
although in this case it should give up due to the use of keyword arguments param, as
in the second example.

```
#!python

def decorator(f):

    def new_func(**kwargs):
        f(junk=kwargs["foo"])

    return new_func

@decorator
def junk1(junk=None):
    print("%s" % junk)

def main():
    junk1(foo=2)
```

---
- Bitbucket: https://bitbucket.org/logilab/pylint/issue/258

**Comment by Pierre-Sassoulas** [comment]
```python
# pylint: disable=missing-docstring

def decorator(func):
    def new_func(internal_arg=3):
        func(junk=internal_arg)

    return new_func


@decorator
def junk1(junk=None):
    print("%s" % junk)


def main():
    junk1(internal_arg=2)
```

Or
```python
# pylint: disable=missing-docstring


def decorator(func):
    def new_func(**kwargs):
        func(junk=kwargs["internal_arg"])

    return new_func


@decorator
def junk1(junk=None):
    print("%s" % junk)


def main():
    junk1(internal_arg=2)

```

Still incorrectly report : ``e.py:15:4: E1123: Unexpected keyword argument 'internal_arg' in function call (unexpected-keyword-arg)``
